### PR TITLE
@types/jsonwebtoken add verify secretOrPublicKey function

### PR DESCRIPTION
**Pull request by anthony-telljohann**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: [https://github.com/auth0/node-jsonwebtoken/blame/master/README.md#L127](https://github.com/auth0/node-jsonwebtoken/blame/master/README.md#L127)
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

**Comment by atelljohannsmothers** [comment]
Not really sure what version to specify... The latest version available on NPM is 7.2.8... so I feel like the version should be updated to 7.2.9.

File: types/jsonwebtoken/index.d.ts [hunk 1]
```diff
@@ -77,6 +78,19 @@ export interface DecodeOptions {
     complete?: boolean;
     json?: boolean;
 }
+
+export interface SecretOrPublicKeyCallback {
+    (
+        err: VerifyErrors,
+        secretOrPublicKey: string | Buffer
+    ): void
```
Comment (SimonSchick) [inline types/jsonwebtoken/index.d.ts]: Missing semicolons.

File: types/jsonwebtoken/jsonwebtoken-tests.ts [hunk 2]
```diff
@@ -110,6 +110,13 @@ jwt.verify(token, cert, { ignoreExpiration: true }, function(err, decoded) {
     // if ignoreExpration == false and token is expired, err == expired token
 });
 
+// verify with get secret or public key function
+jwt.verify(token, (header, callback) => {
+    callback(null, fs.readFileSync("public.pem"))
```
Comment (SimonSchick) [inline types/jsonwebtoken/jsonwebtoken-tests.ts]: See above.

**Review by SimonSchick — CHANGES_REQUESTED** [review]

**Comment by atelljohannsmothers** [comment]
@SimonSchick I added those missing semis.

**Review by SimonSchick — APPROVED** [review]

File: types/jsonwebtoken/index.d.ts [hunk 3]
```diff
@@ -77,6 +78,19 @@ export interface DecodeOptions {
     complete?: boolean;
     json?: boolean;
 }
+
+export interface SecretOrPublicKeyCallback {
```
Comment (SimonSchick) [inline types/jsonwebtoken/index.d.ts]: These should be `type`s not interfaces.

Comment (atelljohannsmothers) [reply types/jsonwebtoken/index.d.ts]: Thanks @SimonSchick, I'll fix that.

Comment (atelljohannsmothers) [reply types/jsonwebtoken/index.d.ts]: @SimonSchick Updated those callback interfaces to callback types.

File: types/jsonwebtoken/index.d.ts [hunk 4]
```diff
@@ -77,6 +78,19 @@ export interface DecodeOptions {
     complete?: boolean;
     json?: boolean;
 }
+
+export type SecretOrPublicKeyCallback = {
```
Comment (SimonSchick) [inline types/jsonwebtoken/index.d.ts]: This is not what I meant, this should be `export type SecretOrPublicKeyCallback  = (err: VerifyErrors, secretOrPublicKey: string | Buffer): void`

File: types/jsonwebtoken/index.d.ts [hunk 5]
```diff
@@ -143,18 +157,18 @@ export declare function verify(
 /**
  * Asynchronously verify given token using a secret or a public key to get a decoded token
  * @param {String} token - JWT string to verify
- * @param {String|Buffer} secretOrPublicKey - Either the secret for HMAC algorithms, or the PEM encoded public key for RSA and ECDSA.
+ * @param {String|Buffer|GetSecretOrPublicKeyCallback} secretOrPublicKey - Either the secret for HMAC algorithms, the PEM encoded public key for RSA and ECDSA, or a function that can fetch the secret or public key.
```
Comment (SimonSchick) [inline types/jsonwebtoken/index.d.ts]: You can delete all jsdoc types, they are redundant.

File: types/jsonwebtoken/index.d.ts [hunk 6]
```diff
@@ -77,6 +78,19 @@ export interface DecodeOptions {
...
+
+export type SecretOrPublicKeyCallback = {
+    (
+        err: VerifyErrors,
+        secretOrPublicKey: string | Buffer
+    ): void;
+}
+export type GetSecretOrPublicKeyCallback = {
+    (
+        header: any,
```
Comment (SimonSchick) [inline types/jsonwebtoken/index.d.ts]: Sorry to drag this out but `any` is not very precise, this is always going to be an `object`, you could even go as far as documenting the header here as: `export interface Header { alg: string, typ?: string }`.

**Review by SimonSchick — CHANGES_REQUESTED** [review]

File: types/jsonwebtoken/index.d.ts (hunk 6 above)
Comment (atelljohannsmothers) [reply types/jsonwebtoken/index.d.ts]: I can add that. I'm just concerned about not having all of the possible JWT header claims here. If I could find an exhaustive list, I could add them all.

File: types/jsonwebtoken/index.d.ts (hunk 4 above)
Comment (atelljohannsmothers) [reply types/jsonwebtoken/index.d.ts]: My bad! I'll fix that right now.

**Comment by atelljohannsmothers** [comment]
@SimonSchick I think I addressed all of your concerns. Let me know if there's anything else you would like me to modify. 👍

**Review by SimonSchick — APPROVED** [review]

**Comment by rlgod** [comment]
@atelljohannsmothers This might have already been fixed in my PR that was just approved this past weekend. Can you confirm? https://github.com/DefinitelyTyped/DefinitelyTyped/pull/29908

**Comment by atelljohannsmothers** [comment]
It was, closing now.

**Pull request closed without merging**